pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java

```diff
@@ -102,79 +122,101 @@ public TransactionBufferHandlerImpl(PulsarClient pulsarClient,
+                            onResponse(peek);
+                        }
+                    }, operationTimeoutInMills, TimeUnit.MILLISECONDS);
+                    op.cmd.retain();
+                    clientCnx.ctx().writeAndFlush(op.cmd, clientCnx.ctx().voidPromise());
+                } else {
+                    invalidateLookupCache(op);
+                    op.cb.completeExceptionally(
+                            new PulsarClientException.LookupException(op.topic + " endTxn channel is not active"));
+                    onResponse(op);
+                }
+            } else {
+                log.error("endTxn error topic: [{}]", op.topic, throwable);
+                invalidateLookupCache(op);
+                op.cb.completeExceptionally(
+                        new PulsarClientException.LookupException(throwable.getMessage()));
+                onResponse(op);
+            }
+        });
+    }
+
     @Override
     public void handleEndTxnOnTopicResponse(long requestId, CommandEndTxnOnPartitionResponse response) {
-        OpRequestSend op = pendingRequests.remove(requestId);
+        OpRequestSend op = outstandingRequests.remove(requestId);
         if (op == null) {
             if (log.isDebugEnabled()) {
                 log.debug("Got end txn on topic response for timeout {} - {}", response.getTxnidMostBits(),
                         response.getTxnidLeastBits());
             }
             return;
         }
-
-        if (!response.hasError()) {
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] Got end txn on topic response for for request {}", op.topic, response.getRequestId());
+        try {
+            if (!response.hasError()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Got end txn on topic response for for request {}", op.topic,
+                            response.getRequestId());
+                }
+                op.cb.complete(new TxnID(response.getTxnidMostBits(), response.getTxnidLeastBits()));
+            } else {
+                log.error("[{}] Got end txn on topic response for request {} error {}", op.topic,
+                        response.getRequestId(),
+                        response.getError());
+                invalidateLookupCache(op);
+                op.cb.completeExceptionally(ClientCnx.getPulsarClientException(response.getError(),
+                        response.getMessage()));
             }
-            op.cb.complete(new TxnID(response.getTxnidMostBits(), response.getTxnidLeastBits()));
-        } else {
-            log.error("[{}] Got end txn on topic response for request {} error {}", op.topic, response.getRequestId(),
-                    response.getError());
-            cache.invalidate(op.topic);
-            op.cb.completeExceptionally(ClientCnx.getPulsarClientException(response.getError(), response.getMessage()));
+        } catch (Exception e) {
```

Comment:
does we need this try catch?

Reply:
Just want to add an error log if there are some exceptions when completing the callback, the final block ensure the onResponse() can be complete.